### PR TITLE
Provider: Constant-time RSA padding, implicit rejection, RSA blinding 

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

File: src/provider/constant_time.h [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+/* Convenience method for signed integers. */
+static ossl_inline int constant_time_select_int(unsigned int mask, int a,
+                                                int b);
+
+
+static ossl_inline unsigned int constant_time_msb(unsigned int a)
+{
+    return 0 - (a >> (sizeof(a) * 8 - 1));
+}
+
```
Comment (jschmidb) [inline src/provider/constant_time.h:108]: remove one blank line?

File: src/provider/rsa_padding.c [hunk 2]
```diff
@@ -669,50 +685,77 @@ int ibmca_rsa_check_pkcs1_tls_padding(const struct ibmca_prov_ctx *provctx,
...
 
-    *outlen = IBMCA_SSL_MAX_MASTER_KEY_LENGTH;
-    if (out != NULL)
-        memcpy(out, &in[inlen - IBMCA_SSL_MAX_MASTER_KEY_LENGTH],
-               IBMCA_SSL_MAX_MASTER_KEY_LENGTH);
+    ok &= version_ok;
 
-    ibmca_debug_ctx(provctx, "outlen: %lu", *outlen);
+    /*
+     * Now copy the result over to the to buffer if good, or random data if
```
Comment (jschmidb) [inline src/provider/rsa_padding.c:1043]: over to the buffer ?

File: src/provider/rsa_blinding.c [hunk 3]
```diff
@@ -52,7 +132,92 @@ static BN_BLINDING *ibmca_rsa_setup_blinding(struct ibmca_key *key)
 
     BN_set_flags(n, BN_FLG_CONSTTIME);
 
-    blinding = BN_BLINDING_create_param(NULL, e, n, bn_ctx, NULL, NULL);
+    /*
+     * Setup the BN_MONT_CTX if needed, it is required by for the mod-expo
+     * callback passed to BN_BLINDING_create_param(). The callback won't be
+     * called if BN_MONT_CTX is NULL.
+     * We hold the wrote lock on blinding_lock when this function is called,
```
Comment (jschmidb) [inline src/provider/rsa_blinding.c:139]: write lock

File: src/provider/rsa_padding.c [hunk 4]
```diff
@@ -524,116 +812,127 @@ int ibmca_rsa_check_oaep_mgf1_padding(const struct ibmca_prov_ctx *provctx,
                     inlen, outsize, EVP_MD_get0_name(oaep_md),
                     EVP_MD_get0_name(mgf1_md), label_len);
 
-    oaep_md_len = EVP_MD_get_size(oaep_md);
-    if (oaep_md_len <= 0) {
+    mdlen = EVP_MD_get_size(oaep_md);
+    if (mdlen <= 0) {
```
Comment (jschmidb) [inline src/provider/rsa_padding.c:816]: unsigned mdlen is never < 0, no need to check

File: src/provider/rsa_padding.c (hunk 2 above)
Comment (ifranzki) [reply src/provider/rsa_padding.c:1043]: fixed

File: src/provider/rsa_padding.c (hunk 4 above)
Comment (ifranzki) [reply src/provider/rsa_padding.c:816]: in deed. removed the check.

File: src/provider/rsa_padding.c [hunk 5]
```diff
@@ -318,6 +459,58 @@ int ibmca_rsa_check_pkcs1_padding_type2(const struct ibmca_prov_ctx *provctx,
...
+            goto out;
+
+        /* decide how long the random message should be */
+        if (ibmca_rsa_prf(provctx, candidate_lengths, sizeof(candidate_lengths),
+                         "length", 6, kdk, kdklen,
+                         MAX_LEN_GEN_TRIES * sizeof(len_candidate) * 8) != 1)
+            goto out;
+
+        /*
+         * max message size is the size of the modulus size less 2 bytes for
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply src/provider/rsa_padding.c:486]: fixed, although this was copied that way from OpenSSL code.....

File: src/provider/rsa_blinding.c (hunk 3 above)
Comment (ifranzki) [reply src/provider/rsa_blinding.c:139]: fixed

**Review by jschmidb — APPROVED** [review]

**Comment by ifranzki** [comment]
close/reopen to trigger travis

**Pull request merged**